src/data.rs

```diff
@@ -60,6 +67,24 @@ impl Data {
             Data::PrivAppendable(ref data) => data.identifier(),
         }
     }
+
+    /// Validate data size.
+    pub fn validate_size(&self) -> bool {
+        let serialised_data = match serialisation::serialise(&self) {
```

Comment:
This will cause a lot of duplicate serialisation; we should consider re-adding the `payload` getter to (some of) the data types and check that first here, so that e.g. an immutable data with a 2MB payload doesn't get serialised before being rejected.
(Maybe just add a `TODO` here for now.)